### PR TITLE
WINC-950: Test WMCO version upgrades

### PR DESCRIPTION
**Pull request by sebsoto**
Adds the upgrade-test-setup and upgrade-test make targets, which are
designed to be ran before a WMCO upgrade occurs and after,
respectively. These targets will do basic testing on WMCO to ensure that
WMCO is able to upgrade nodes.

### CODE REVIEW

File: test/e2e/create_test.go [hunk 1]
```diff
@@ -42,8 +42,8 @@ const (
 func creationTestSuite(t *testing.T) {
 	tc, err := NewTestContext()
 	require.NoError(t, err)
-	// The order of tests here are important. Any node object related tests should be run only after
-	// testWindowsNodeCreation as that initializes the node objects in the global context.
+
+	require.NoError(t, tc.loadExistingNodes(), "error getting the current Windows nodes in the cluster")
 	if !t.Run("Creation", tc.testWindowsNodeCreation) {
```
Comment (aravindhp) [inline test/e2e/create_test.go:47]: Should we separate this into its own test suite and make this a validationTestSuite()?

**Review by aravindhp — APPROVED** [review]
This is great work, @sebsoto. I just have one question but I am approving either way to move things forward.

File: pkg/daemon/controller/controller.go [hunk 2]
```diff
@@ -252,19 +270,50 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
...
+// handleReconcileFailure should be ran when a reconciliation fails due to issues on the Windows instance
+func (sc *ServiceController) handleReconcileFailure(node *core.Node) error {
+	sc.failureCount += 1
+	if sc.failureCount < 5 {
+		return nil
+	}
+	if err := drain.RunCordonOrUncordon(sc.drainer, node, true); err != nil {
+		return err
+	}
+	return nil
```
Comment (saifshaikh48) [inline pkg/daemon/controller/controller.go:300]: condense to `return drain.RunCordonOrUncordon(sc.drainer, node, true)`

File: test/e2e/secrets_test.go
Comment (aravindhp) [inline test/e2e/secrets_test.go:1]: What does this have to do with upgrade testing? IOW, if you want to include this commit in the upgrade PR please update the commit message with how it is helping with the upgrade testing.

File: test/e2e/create_test.go
Comment (aravindhp) [inline test/e2e/create_test.go:1]: What does this have to do with upgrade testing?

File: test/e2e/create_test.go [hunk 3]
```diff
@@ -13,23 +13,19 @@ import (
...
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/wait"
-	cloudproviderapi "k8s.io/cloud-provider/api"
-
-	"github.com/openshift/windows-machine-config-operator/controllers"
```
Comment (aravindhp) [inline test/e2e/create_test.go:25]: Don't we keep the WMCO packages in a separate stanza?

File: test/e2e/network_test.go [hunk 4]
```diff
@@ -29,9 +29,9 @@ import (
 func testNetwork(t *testing.T) {
 	tc, err := NewTestContext()
 	require.NoError(t, err)
-	err = tc.waitForWindowsNodes(gc.numberOfMachineNodes, false, false, false)
+	err = tc.waitForConfiguredWindowsNodes(gc.numberOfMachineNodes, false, false)
```
Comment (aravindhp) [inline test/e2e/network_test.go:32]: This test needs a node that is schedulable. How are you guaranteeing that now that you have moved to just checking if the node has been configured? This becomes especially important if I am just running the networking test without running the create tests but bringing up nodes outside of the test suite.

File: hack/run-ci-e2e-test.sh [hunk 5]
```diff
@@ -140,12 +140,15 @@ GO_TEST_ARGS="$BYOH_NODE_COUNT_OPTION $MACHINE_NODE_COUNT_OPTION --private-key-p
...
+  # Run the creation tests of the Windows VMs
+  printf "\n####### Testing creation #######\n" >> "$ARTIFACT_DIR"/wmco.log
+  go test ./test/e2e/... -run=TestWMCO/create -v -timeout=90m -args $GO_TEST_ARGS
+  # Get logs for the creation tests
+  printf "\n####### WMCO logs for creation tests #######\n" >> "$ARTIFACT_DIR"/wmco.log
+  get_WMCO_logs
+fi
 
 if [[ "$TEST" = "all" || "$TEST" = "basic" ]]; then
   printf "\n####### Testing network #######\n" >> "$ARTIFACT_DIR"/wmco.log
```
Comment (aravindhp) [inline hack/run-ci-e2e-test.sh:154]: Shouldn't you doing network tests as part of the upgrade suite?

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: test/e2e/network_test.go (hunk 4 above)
Comment (saifshaikh48) [reply test/e2e/network_test.go:32]: +1, might have to run testNodesBecomeReady() here instead

File: test/e2e/reconfigure_test.go
Comment (saifshaikh48) [inline test/e2e/reconfigure_test.go:1]: > WMCO's controllers will return early, and not change the condition state.

Not sure I understand this. Every WMCO controller changes the upgradeable condition state when its Reconcile function returns without error (via a defer call).

File: test/e2e/network_test.go (hunk 4 above)
Comment (sebsoto) [reply test/e2e/network_test.go:32]: > How are you guaranteeing that now

I am not sure why this needs to be guaranteed within this test?
The test itself will fail if a node is not schedulable. This is the networking test, retesting node statuses that we test within the creation/validation tests is overlapping responsibilities.

File: test/e2e/create_test.go [hunk 6]
```diff
@@ -471,33 +471,6 @@ func (tc *testContext) waitForWindowsNodes(nodeCount int32, expectError, checkVe
 		}
 
 		for _, node := range nodes {
-			// check node status
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply test/e2e/create_test.go:474]: Ready != schedulable

File: hack/run-ci-e2e-test.sh (hunk 5 above)
Comment (sebsoto) [reply hack/run-ci-e2e-test.sh:154]: Maybe? There needs to be a followup WMCO PR to dial in the tests, after the release job is created. In a way this PR is just creating the stubs. I will make this clear in the commit message, and perhaps we can workout out the specifics of this test in that follow-up.

File: test/e2e/network_test.go (hunk 4 above)
Comment (sebsoto) [reply test/e2e/network_test.go:32]: To elaborate, schedulable is a part of the node spec that is applied by WMCO or a user. IMO checking for that would be similar to checking for any taint that isn't tolerated.

File: test/e2e/secrets_test.go
Comment (sebsoto) [reply test/e2e/secrets_test.go:1]: Commit message updated

File: test/e2e/create_test.go
Comment (sebsoto) [reply test/e2e/create_test.go:1]: Commit message updated

File: test/e2e/reconfigure_test.go
Comment (saifshaikh48) [reply test/e2e/reconfigure_test.go:1]: I see, in that case there is time where WMCO is not busy but WICD is. Still, wouldn't this only affect the first check? You are removing both
If you saw both checks causing issues it's fine, I'm just trying to understand why

File: test/e2e/secrets_test.go [hunk 7]
```diff
@@ -95,17 +95,17 @@ func (tc *testContext) testUserDataTamper(t *testing.T) {
...
 	for _, node := range gc.machineNodes {
 		oldNodes = append(oldNodes, node.GetName())
 	}
 
-	// waitForWindowsNodes will re-populate gc.machineNodes with the configured nodes found
+	// waitForConfiguredWindowsNodes will re-populate gc.machineNodes with the configured nodes found
 	log.Printf("waiting for existing Machine nodes to be removed and replaced")
 	return wait.Poll(retryInterval, time.Minute*10, func() (done bool, err error) {
-		err = tc.waitForWindowsNodes(gc.numberOfMachineNodes, false, false, false)
+		err = tc.waitForConfiguredWindowsNodes(gc.numberOfMachineNodes, false, false)
```
Comment (saifshaikh48) [inline test/e2e/secrets_test.go:108]: we no longer ensure the nodes become schedulable after private key change. is this intended?

File: test/e2e/create_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply test/e2e/create_test.go:474]: I think you've renamed the function since this comment was originally left. The point still stands. We need to check for ready AND schedulable in more than just the initial create suite.

File: test/e2e/network_test.go (hunk 4 above)
Comment (aravindhp) [reply test/e2e/network_test.go:32]: >I am not sure why this needs to be guaranteed within this test?
The test itself will fail if a node is not schedulable. This is the networking test, retesting node statuses that we test within the creation/validation tests is overlapping responsibilities.

You are assuming the creation test suite was run prior. What if I brought up the nodes manually and then manually ran the just networking test suite? Will `loadExistingNodes()` take care of that?

File: test/e2e/secrets_test.go (hunk 7 above)
Comment (aravindhp) [reply test/e2e/secrets_test.go:108]: +1

File: test/e2e/reconfigure_test.go [hunk 8]
```diff
@@ -46,10 +46,6 @@ func (tc *testContext) reconfigurationTest(t *testing.T) {
 	_, err = tc.client.K8s.CoreV1().Nodes().Patch(context.TODO(), machineNodes[0].Name, types.JSONPatchType,
 		patchData, metav1.PatchOptions{})
 	require.NoError(t, err)
-	// Ensure operator communicates to OLM that upgrade is not safe when processing Machine nodes
-	err = tc.validateUpgradeableCondition(metav1.ConditionFalse)
```
Comment (aravindhp) [inline test/e2e/reconfigure_test.go:50]: > If WICD fixes the change by reconciling the node, before WMCO is able to reconcile the Machine or windows instance ConfigMap, WMCO's controllers will return early, and not change the condition state.

I don't follow the commit message. Isn't this a bug then? I guess I am not understanding the timing issue here.

File: hack/run-ci-e2e-test.sh (hunk 5 above)
Comment (aravindhp) [reply hack/run-ci-e2e-test.sh:154]: Weren't we running networking tests as part of  the current "upgrade" test? So why is this a maybe?

File: test/e2e/network_test.go (hunk 4 above)
Comment (sebsoto) [reply test/e2e/network_test.go:32]: This is still being tested, its just not a separate check. Specifically, it will be tested by the failure of the windows server deployments to be scheduled on the node. The pod events will state that no nodes meet the scheduling criteria.

File: test/e2e/reconfigure_test.go
Comment (sebsoto) [reply test/e2e/reconfigure_test.go:1]: Moved to different PR

File: test/e2e/secrets_test.go (hunk 7 above)
Comment (sebsoto) [reply test/e2e/secrets_test.go:108]: Added the check below

File: test/e2e/reconfigure_test.go (hunk 8 above)
Comment (sebsoto) [reply test/e2e/reconfigure_test.go]: Moved to separate PR

File: test/e2e/validation_test.go [hunk 9]
```diff
@@ -53,6 +54,53 @@ type winService struct {
 	description string
 }
 
+// testNodesBecomeReady tests that all Windows nodes become ready and schedulable at least once.
```
Comment (aravindhp) [inline test/e2e/validation_test.go:57]: I don't understand the `at least once` comment. Why is that there?

File: test/e2e/validation_test.go [hunk 10]
```diff
@@ -53,6 +54,53 @@ type winService struct {
...
+func (tc *testContext) testNodesBecomeReady(t *testing.T) {
+	nodes := gc.allNodes()
+	for index, node := range nodes {
+		t.Run(node.GetName(), func(t *testing.T) {
+			err := wait.PollImmediate(retry.Interval, retry.ResourceChangeTimeout, func() (done bool, err error) {
+				// Get current node state, and update it in the cache
```
Comment (aravindhp) [inline test/e2e/validation_test.go:63]: What cache are you updating? Isn't nodes a local variable?

If you are updating the cache it shouldn't be a side effect of testing if nodes are ready.

File: test/e2e/validation_test.go [hunk 11]
```diff
@@ -53,6 +54,53 @@ type winService struct {
...
+				foundNode, err := tc.client.K8s.CoreV1().Nodes().Get(context.TODO(), node.GetName(), meta.GetOptions{})
+				require.NoError(t, err)
+				nodes[index] = *foundNode
+				return tc.nodeReadyAndSchedulable(*foundNode), nil
```
Comment (aravindhp) [inline test/e2e/validation_test.go:67]: This implies, the function should be name `testNodesBecomeReadyAndSchedulable()`. Or am I missing something?

File: test/e2e/validation_test.go [hunk 12]
```diff
@@ -53,6 +54,53 @@ type winService struct {
...
+			})
+			assert.NoError(t, err)
+		})
+	}
+}
+
+func (tc *testContext) nodeReadyAndSchedulable(node core.Node) bool {
```
Comment (aravindhp) [inline test/e2e/validation_test.go:74]: Please add a doc

File: test/e2e/validation_test.go [hunk 13]
```diff
@@ -742,13 +790,69 @@ func (tc *testContext) getOperatorConditionName() (string, error) {
...
+
+			pubKey, err := tc.pubKeyAnnotationCorrect(&node)
+			require.NoError(t, err)
+			assert.True(t, pubKey)
+		})
+	}
+}
+
+// usernameAnnotationCorrect checks that the username annotation value is decipherable and correct
+func (tc *testContext) usernameAnnotationCorrect(node *core.Node) (bool, error) {
```
Comment (aravindhp) [inline test/e2e/validation_test.go:815]: If you want to go with this sort of naming I suggest `isUsernameAnnotationCorrect()` though I prefer `checkUserNameAnnotation()`

File: test/e2e/validation_test.go [hunk 14]
```diff
@@ -742,13 +790,69 @@ func (tc *testContext) getOperatorConditionName() (string, error) {
...
+		return false, err
+	}
+	if username != tc.vmUsername() {
+		return false, nil
+	}
+	return true, nil
+}
+
+// pubKeyAnnotationCorrect that node is annotated with the public key which matches the private key used to configure it
+func (tc *testContext) pubKeyAnnotationCorrect(node *core.Node) (bool, error) {
```
Comment (aravindhp) [inline test/e2e/validation_test.go:836]: Same comment applies to the name here.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: test/e2e/validation_test.go (hunk 9 above)
Comment (sebsoto) [reply test/e2e/validation_test.go]: Thought process is that schedulable can be changed by WMCO at will. Removing it for clarity

File: test/e2e/validation_test.go (hunk 10 above)
Comment (sebsoto) [reply test/e2e/validation_test.go]: gc.Nodes is not a local variable. I agree that it shouldn't be a side effect of testing, but thats already the case with waitForConfiguredWindowsNodes. A large portion of the tests depend on there being an updated node cache available to them via gc, which is why waitForConfiguredNodes have been called at various points in the suite.

I will remove the caching, I don't think it will cause any issues. Being this defensive was an artifact of the cancelled work to move cordoning into WICD.

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

Thanks for working on this. e2e tests have been nicely refactored too

**Review by aravindhp — APPROVED** [review]
Nice work, @sebsoto
